mobility/choreographer/test/test_zones_keepout.cc

```diff
@@ -0,0 +1,158 @@
+ * 
+ * All rights reserved.
+ * 
+ * The Astrobee platform is licensed under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+// Required for the test framework
+#include <gtest/gtest.h>
+
+// Required for the test cases
+#include <ros/ros.h>
+#include <ros/console.h>
+#include <actionlib/client/simple_action_client.h>
+
+// Listen for transforms
+#include <tf2_ros/transform_listener.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+// FSW includes
+#include <ff_util/ff_names.h>
+#include <ff_util/ff_action.h>
+#include <ff_util/config_client.h>
+
+// FF messages
+#include <ff_msgs/MotionAction.h>
+#include <ff_msgs/EkfState.h>
+
+// C++ STL includes
+#include <iostream>
+#include <memory>
+
+// Avoid sending the command multiple times
+bool sent_ = false;
+bool stable_ = false;
+
+// Subscriber Ekf
+ros::Subscriber sub_ekf_;
+  ff_util::FreeFlyerActionClient<ff_msgs::MotionAction> client_t_;
+
+tf2_ros::Buffer tf_buffer_;
+
+// Called once when the goal completes
+void MResultCallback(ff_util::FreeFlyerActionState::Enum result_code,
+  ff_msgs::MotionResultConstPtr const& result) {
+  EXPECT_EQ(result->response, ff_msgs::MotionResult::VIOLATES_KEEP_OUT);
+  ROS_ERROR("Test Passed");
+  ros::shutdown();
+}
+
+// Called every time feedback is received for the goal
+void MFeedbackCallback(ff_msgs::MotionFeedbackConstPtr const& feedback) {}
+
+
+
+void StateCallback(const ff_msgs::EkfStateConstPtr& state) {
+  if (!client_t_.IsConnected()) return;
+  else if (state->confidence != ff_msgs::EkfState::CONFIDENCE_GOOD) return;
+  else
+    sub_ekf_.shutdown();
+  sleep(1);
+  stable_ = true;
+  return;
+}
+
+// Keepout zone test
+TEST(choreographer_nominal, ZoneBreach) {
+  // Get the node handle
+  ros::NodeHandle nh;
+
+  tf2_ros::TransformListener tfListener(tf_buffer_);
+
+  // Setup MOBILITY action
+  client_t_.SetConnectedTimeout(30.0);
+  client_t_.SetActiveTimeout(30.0);
+  client_t_.SetResponseTimeout(30.0);
+  client_t_.SetFeedbackCallback(std::bind(
+    MFeedbackCallback, std::placeholders::_1));
+  client_t_.SetResultCallback(std::bind(
+    MResultCallback, std::placeholders::_1, std::placeholders::_2));
+  client_t_.Create(&nh, ACTION_MOBILITY_MOTION);
+
+  // Wait for Ekf to start publishing to send the motion goal
+  sub_ekf_ = nh.subscribe(TOPIC_GNC_EKF, 1000, &StateCallback);
+
+  // This waits until the simulation is in a stable status
+  while (!stable_) ros::spinOnce();  // Mobility
```

Comment:
Is there anyway this would cause an infinite loop? You could add a counter to the state callback and fail the test if the counter is greater than 30 seconds of ekf messages.

Reply:
The rostest has a timeout of 60seconds, meaning that if the test is not finished within 60 seconds, then it fails